pennylane_lightning/src/gates/cpu_kernels/GateImplementationsPI.hpp

```diff
@@ -1059,4 +1059,285 @@ class GateImplementationsPI : public PauliGenerator<GateImplementationsPI> {
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+extern template auto PauliGenerator<GateImplementationsPI>::applyGeneratorRZ(
+    std::complex<double> *, size_t, const std::vector<size_t> &, bool)
+    -> double;
+
+extern template auto GateImplementationsPI::applyGeneratorIsingXX(
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+extern template auto
+GateImplementationsPI::applyGeneratorIsingXX(std::complex<double> *, size_t,
+                                             const std::vector<size_t> &, bool)
+    -> double;
+
+extern template auto
+GateImplementationsPI::applyGeneratorIsingXY(std::complex<double> *, size_t,
+                                             const std::vector<size_t> &, bool)
+    -> double;
+extern template auto GateImplementationsPI::applyGeneratorIsingXY(
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+
+extern template auto GateImplementationsPI::applyGeneratorIsingYY(
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+extern template auto
+GateImplementationsPI::applyGeneratorIsingYY(std::complex<double> *, size_t,
+                                             const std::vector<size_t> &, bool)
+    -> double;
+
+extern template auto
+GateImplementationsPI::applyGeneratorIsingZZ(std::complex<double> *, size_t,
+                                             const std::vector<size_t> &, bool)
+    -> double;
+extern template auto GateImplementationsPI::applyGeneratorIsingZZ(
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+
+extern template auto GateImplementationsPI::applyGeneratorCRX(
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+extern template auto
+GateImplementationsPI::applyGeneratorCRX(std::complex<double> *, size_t,
+                                         const std::vector<size_t> &, bool)
+    -> double;
+
+extern template auto GateImplementationsPI::applyGeneratorCRY(
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+extern template auto
+GateImplementationsPI::applyGeneratorCRY(std::complex<double> *, size_t,
+                                         const std::vector<size_t> &, bool)
+    -> double;
+
+extern template auto GateImplementationsPI::applyGeneratorCRZ(
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+extern template auto
+GateImplementationsPI::applyGeneratorCRZ(std::complex<double> *, size_t,
+                                         const std::vector<size_t> &, bool)
+    -> double;
+
+extern template auto GateImplementationsPI::applyGeneratorControlledPhaseShift(
+    std::complex<float> *, size_t, const std::vector<size_t> &, bool) -> float;
+extern template auto GateImplementationsPI::applyGeneratorControlledPhaseShift(
+    std::complex<double> *, size_t, const std::vector<size_t> &, bool)
+    -> double;
```

Comment:
As above with potentially moving these into a separate file for inclusion here.

Reply:
As `extern` explicitly force the compiler not to instantiate the functions, I concern that separating these statements to another file fails to achieve this objective. One happens to include only the class definition and does not aware he/she made another instantiation. Unfortunately, I couldn't find a way to make the compiler warns or raises an error when this happens (besides moving all member function definitions to cpp files, only remaining declarations).

Reply:
Another solution I can think is one from this comment: https://github.com/PennyLaneAI/pennylane-lightning/pull/324#issuecomment-1200073065

Reply:
Hmm, good point. Since we need this to work now, probably best to just power ahead and keep this for now. That ok with you? We can revisit then potentially later once the new kernels are through.

Reply:
:+1:

Reply:
Great, nothing else to add from my side. If you are happy too, I will approve.